### ISSUE TITLE
[Bug]: The datanode panic occurred in the chaos test of all components with error `invalid memory address or nil pointer dereference `

### ISSUE THREAD

**Issue by zhuwenxing** [milvus-io/milvus#37568]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version:master-20241108-e073906a-amd64
- Deployment mode(standalone or cluster):all
- MQ type(rocksmq, pulsar or kafka):    
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

```

[2024/11/08 08:32:28.968 +00:00] [DEBUG] [config/etcd_source.go:154] ["etcd refreshConfigurations"] [prefix=by-dev/config] [endpoints="[etcd-followers-pod-kill-19402-0.etcd-followers-pod-kill-19402-headless.chaos-testing.svc.cluster.local:2379,etcd-followers-pod-kill-19402-1.etcd-followers-pod-kill-19402-headless.chaos-testing.svc.cluster.local:2379,etcd-followers-pod-kill-19402-2.etcd-followers-pod-kill-19402-headless.chaos-testing.svc.cluster.local:2379]"]
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x41d2535]

goroutine 25234 gp=0xc001edba40 m=9 mp=0xc00088a008 [running]:
panic({0x62743a0?, 0x9c25020?})
	/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/runtime/panic.go:779 +0x158 fp=0xc00b605fd8 sp=0xc00b605f28 pc=0x1f92b18
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.deferwrap1()
	/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.28.0/trace/span.go:398 +0x25 fp=0xc00b605ff8 sp=0xc00b605fd8 pc=0x2a88625
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc0013991e0, {0x0, 0x0, 0x0?})
	/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.28.0/trace/span.go:436 +0xa82 fp=0xc00b606460 sp=0xc00b605ff8 pc=0x2a885a2
github.com/milvus-io/milvus/internal/datanode/compaction.(*mixCompactionTask).Compact.deferwrap1()
	/workspace/source/internal/datanode/compaction/mix_compactor.go:297 +0x2c fp=0xc00b606490 sp=0xc00b606460 pc=0x49c06ec
panic({0x62743a0?, 0x9c25020?})
	/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/runtime/panic.go:770 +0x132 fp=0xc00b606540 sp=0xc00b606490 pc=0x1f92af2
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.deferwrap1()
	/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.28.0/trace/span.go:398 +0x25 fp=0xc00b606560 sp=0xc00b606540 pc=0x2a88625
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc001399380, {0x0, 0x0, 0xc007fca900?})
	/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.28.0/trace/span.go:436 +0xa82 fp=0xc00b6069c8 sp=0xc00b606560 pc=0x2a885a2
github.com/milvus-io/milvus/internal/datanode/compaction.(*mixCompactionTask).mergeSplit.deferwrap1()
	/workspace/source/internal/datanode/compaction/mix_compactor.go:141 +0x2c fp=0xc00b6069f8 sp=0xc00b6069c8 pc=0x49bdc6c
panic({0x62743a0?, 0x9c25020?})
	/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/runtime/panic.go:770 +0x132 fp=0xc00b606aa8 sp=0xc00b6069f8 pc=0x1f92af2
runtime.panicmem(...)
	/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/runtime/panic.go:261
runtime.sigpanic()
	/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/runtime/signal_unix.go:881 +0x378 fp=0xc00b606b08 sp=0xc00b606aa8 pc=0x1fad818
github.com/apache/arrow/go/v12/arrow/memory.(*Buffer).Len(...)
	/go/pkg/mod/github.com/milvus-io/arrow/go/v12@v12.0.1/arrow/memory/buffer.go:100
github.com/milvus-io/milvus/internal/storage.calculateArraySize.func1(...)
	/workspace/source/internal/storage/serde.go:618
github.com/samber/lo.SumBy[...](...)
	/go/pkg/mod/github.com/samber/lo@v1.27.0/math.go:66
github.com/milvus-io/milvus/internal/storage.calculateArraySize({0x7299800?, 0xc00dfb0640?})
	/workspace/source/internal/storage/serde.go:617 +0x35 fp=0xc00b606b20 sp=0xc00b606b08 pc=0x41d2535
github.com/milvus-io/milvus/internal/storage.(*CompositeRecordWriter).Write(0xc00c9c4dd0, {0x72613d0, 0xc00c9c45f0})
	/workspace/source/internal/storage/serde.go:656 +0x193 fp=0xc00b606bf8 sp=0xc00b606b20 pc=0x41d2833
github.com/milvus-io/milvus/internal/storage.(*SerializeWriter[...]).WriteRecord(0xc00b606cf8, {0x72613d0?, 0xc00c9c45f0})
	/workspace/source/internal/storage/serde.go:838 +0x63 fp=0xc00b606c90 sp=0xc00b606bf8 pc=0x41e3303
github.com/milvus-io/milvus/internal/datanode/compaction.(*SegmentWriter).WriteRecord(0xc013ea9580, {0x72613d0, 0xc00c9c45f0})
	/workspace/source/internal/datanode/compaction/segment_writer.go:433 +0x2e5 fp=0xc00b606d80 sp=0xc00b606c90 pc=0x49c3085
github.com/milvus-io/milvus/internal/datanode/compaction.(*MultiSegmentWriter).WriteRecord(0xc00b6076c0, {0x72613d0, 0xc00c9c45f0})
	/workspace/source/internal/datanode/compaction/segment_writer.go:230 +0x5f fp=0xc00b606db0 sp=0xc00b606d80 pc=0x49c269f
github.com/milvus-io/milvus/internal/datanode/compaction.(*mixCompactionTask).writeSegment.func3({0x72613d0?, 0xc007fca8e8?}, 0x64c2b178d200002?, 0xc001e0dc20?)
	/workspace/source/internal/datanode/compaction/mix_compactor.go:232 +0xd7 fp=0xc00b606ec8 sp=0xc00b606db0 pc=0x49bec97
github.com/milvus-io/milvus/internal/datanode/compaction.(*mixCompactionTask).writeSegment(0xc0026d4990, {0x7251508, 0xc00ab3fad0}, {0xc00ab76200, 0x10, 0x10}, {0x9f3fae0, 0x0, 0x0}, 0xc00b6076c0, ...)
	/workspace/source/internal/datanode/compaction/mix_compactor.go:285 +0x951 fp=0xc00b6072d8 sp=0xc00b606ec8 pc=0x49be5f1
github.com/milvus-io/milvus/internal/datanode/compaction.(*mixCompactionTask).mergeSplit(0xc0026d4990, {0x72515b0, 0xc00ab69260}, 0xc00ab3fa40, 0xc00ab3fa10)
	/workspace/source/internal/datanode/compaction/mix_compactor.go:160 +0x768 fp=0xc00b6077f8 sp=0xc00b6072d8 pc=0x49bd428
github.com/milvus-io/milvus/internal/datanode/compaction.(*mixCompactionTask).Compact(0xc0026d4990)
	/workspace/source/internal/datanode/compaction/mix_compactor.go:350 +0xacf fp=0xc00b607dc8 sp=0xc00b6077f8 pc=0x49bfc6f
github.com/milvus-io/milvus/internal/datanode/compaction.(*executor).executeTask(0xc0001e00e0, {0x7276408, 0xc0026d4990})
	/workspace/source/internal/datanode/compaction/executor.go:172 +0x34c fp=0xc00b607f98 sp=0xc00b607dc8 pc=0x49b1ccc
github.com/milvus-io/milvus/internal/datanode/compaction.(*executor).Start.func1()
	/workspace/source/internal/datanode/compaction/executor.go:153 +0x55 fp=0xc00b607fe0 sp=0xc00b607f98 pc=0x49b18f5
runtime.goexit({})
	/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/runtime/asm_amd64.s:1695 +0x1 fp=0xc00b607fe8 sp=0xc00b607fe0 pc=0x1fd1981
created by github.com/milvus-io/milvus/internal/datanode/compaction.(*executor).Start in goroutine 304
	/workspace/source/internal/datanode/compaction/executor.go:151 +0x45
```

```
[2024-11-08T08:27:54.881Z] + kubectl get pods -o wide

[2024-11-08T08:27:54.883Z] + grep etcd-followers-pod-kill-19402

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-0                                   1/1     Running             0                 39m     10.104.24.23    4am-node29   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-1                                   1/1     Running             0                 16m     10.104.21.143   4am-node24   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-2                                   1/1     Running             0                 39m     10.104.26.215   4am-node32   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-milvus-datanode-cc6b8978-hzkvs      0/1     CrashLoopBackOff    10 (62s ago)      39m     10.104.34.107   4am-node37   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-milvus-datanode-cc6b8978-pj8pd      0/1     CrashLoopBackOff    10 (2m16s ago)    39m     10.104.33.58    4am-node36   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-milvus-indexnode-76d6b5c5cc-4js9b   1/1     Running             3 (38m ago)       39m     10.104.1.183    4am-node10   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-milvus-indexnode-76d6b5c5cc-mpqlk   1/1     Running             3 (38m ago)       39m     10.104.19.128   4am-node28   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-milvus-indexnode-76d6b5c5cc-xbs2j   1/1     Running             3 (38m ago)       39m     10.104.9.89     4am-node14   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-milvus-mixcoord-77976fbfbf-f7d5s    1/1     Running             3 (38m ago)       39m     10.104.33.57    4am-node36   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-milvus-proxy-77f4b4fccf-vk289       1/1     Running             3 (38m ago)       39m     10.104.13.115   4am-node16   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-milvus-querynode-866cb94f6c-9r7j6   1/1     Running             3 (38m ago)       39m     10.104.6.10     4am-node13   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-milvus-querynode-866cb94f6c-q9c7z   1/1     Running             3 (38m ago)       39m     10.104.32.98    4am-node39   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-milvus-querynode-866cb94f6c-tsz6t   1/1     Running             3 (38m ago)       39m     10.104.13.116   4am-node16   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-minio-0                             1/1     Running             0                 39m     10.104.26.212   4am-node32   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-minio-1                             1/1     Running             0                 39m     10.104.24.20    4am-node29   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-minio-2                             1/1     Running             0                 39m     10.104.21.130   4am-node24   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-minio-3                             1/1     Running             0                 39m     10.104.18.4     4am-node25   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-pulsar-bookie-0                     1/1     Running             0                 39m     10.104.21.128   4am-node24   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-pulsar-bookie-1                     1/1     Running             0                 39m     10.104.24.22    4am-node29   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-pulsar-bookie-init-lq86k            0/1     Completed           0                 39m     10.104.21.118   4am-node24   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-pulsar-broker-0                     1/1     Running             0                 39m     10.104.21.119   4am-node24   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-pulsar-proxy-0                      1/1     Running             0                 39m     10.104.14.236   4am-node18   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-pulsar-pulsar-init-9dj7f            0/1     Completed           0                 39m     10.104.32.99    4am-node39   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-pulsar-recovery-0                   1/1     Running             0                 39m     10.104.21.120   4am-node24   <none>           <none>

[2024-11-08T08:27:55.138Z] etcd-followers-pod-kill-19402-pulsar-zookeeper-0                  1/1     Running             0                 39m     10.104.26.213   4am-node32   <none>           <none>
```

### Milvus Log

failed job: https://qa-jenkins.milvus.io/blue/organizations/jenkins/chaos-test-cron/detail/chaos-test-cron/19402/pipeline

log:
[artifacts-etcd-followers-pod-kill-19402-server-logs.tar.gz](https://github.com/user-attachments/files/17694474/artifacts-etcd-followers-pod-kill-19402-server-logs.tar.gz)

### Anything else?

Since this issue appears in almost all chaos tests, it doesn't feel like it's caused by a specific component's chaos, but possibly by the concurrent requests from clients in the chaos test.

**Comment by zhuwenxing** [comment]
```
created by github.com/milvus-io/milvus/internal/datanode/compaction.(*executor).Start in goroutine 304
	/workspace/source/internal/datanode/compaction/executor.go:151 +0x45
```

/assign @XuanYang-cn 
Seems like a compaction issue

**Comment by XuanYang-cn** [comment]
![image](https://github.com/user-attachments/assets/a42e1ee9-ca0d-4b4b-975c-3faa80bb52dd)

**Comment by XuanYang-cn** [comment]
https://qa-jenkins.milvus.io/blue/organizations/jenkins/chaos-test-cron/detail/chaos-test-cron/19416/pipeline/356/

**Comment by XuanYang-cn** [comment]
/assign @zhuwenxing 
/unassign
Should be fixed, please help verify

**Comment by zhuwenxing** [comment]
verified and fixed in master-20241112-c1eccce2-amd64
